File: day15.hs
import System.IO(readFile)
import Data.List.Split(splitOn)
import Data.Map(fromList, insert, Map, member, alter, (!))
-- import Debug.Trace ( trace )
-- ebug = flip trace

enumMp :: [Int] -> Map Int Int
enumMp arr = startMp arr 1 (fromList [])

startMp :: [Int] -> Int -> Map Int Int -> Map Int Int
startMp arr i mp = let
  (h:t) = arr
  in
    if null t then mp
    else startMp t (i+1) (insert h i mp) 

solve :: Int -> [Int] -> Int
solve limit nums = let
  mp = enumMp nums
  in
    solveFull limit mp (length nums) (last nums)

solveFull :: Int -> Map Int Int -> Int -> Int -> Int
solveFull limit mp c current
  | c == limit              = current
  | not (member current mp) = solveFull limit (insert current c mp) (c + 1) 0
  | otherwise               = solveFull limit (alter (\ _ -> Just c) current mp) (c + 1) (c - (mp ! current))

main :: IO()
main = do
  fl <- readFile "inputs/inp_day15.txt"
  let ns = map read $ splitOn "," (init fl)::[Int]
  print $ solve 30000000 ns

File: day15.scala
#!/usr/bin/env scala

import scala.io.Source

val nums = Source.fromFile("inputs/inp_day15.txt").getLines.mkString.split(",").map((s:String) => s.toInt).toList

def solve(limit: Int, nums: List[Int]) : Int = {
  
  var n = 0
  var tmp = 0
  val arr = Array.fill(limit)(0)
  
  for (i <- 1 to nums.length) {
    arr(nums(i - 1)) = i
  }
  
  n = nums.last
  for (i <- nums.length to limit - 1) {
    tmp = arr(n)
    arr(n) = i
    n = if (tmp == 0) 0 else i - tmp
  }

  return n
}

// part 1
// solve(nums.last, 2020, nums diff List(nums.last))

// part 2
val p2 = () => solve(30000000, nums)
// println(p2())